sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
+        AsynchronousFileChannel file, long chunkNum, ParallelTransferOptions optionsReal, Lock progressLock,
+        AtomicLong totalProgress) {
+
+        // Extract the body.
+        Flux<ByteBuffer> data = response.getValue();
+
+        // Report progress as necessary.
+        data = ProgressReporter.addParallelProgressReporting(data,
+            optionsReal.getProgressReceiver(), progressLock, totalProgress);
+
+        // Construct the desired return type.
+        BlobProperties properties = new BlobProperties(null, response.getDeserializedHeaders().getLastModified(),
+            response.getDeserializedHeaders().getETag(),
+            response.getDeserializedHeaders().getContentLength() == null
+                ? 0 : response.getDeserializedHeaders().getContentLength(),
+            response.getDeserializedHeaders().getContentType(), null,
+            response.getDeserializedHeaders().getContentEncoding(),
+            response.getDeserializedHeaders().getContentDisposition(),
+            response.getDeserializedHeaders().getContentLanguage(), response.getDeserializedHeaders().getCacheControl(),
+            response.getDeserializedHeaders().getBlobSequenceNumber(), response.getDeserializedHeaders().getBlobType(),
+            response.getDeserializedHeaders().getLeaseStatus(), response.getDeserializedHeaders().getLeaseState(),
+            response.getDeserializedHeaders().getLeaseDuration(), response.getDeserializedHeaders().getCopyId(),
+            response.getDeserializedHeaders().getCopyStatus(), response.getDeserializedHeaders().getCopySource(),
+            response.getDeserializedHeaders().getCopyProgress(),
+            response.getDeserializedHeaders().getCopyCompletionTime(),
+            response.getDeserializedHeaders().getCopyStatusDescription(),
+            response.getDeserializedHeaders().isServerEncrypted(), null, null, null, null, null,
+            response.getDeserializedHeaders().getEncryptionKeySha256(), null,
+            response.getDeserializedHeaders().getMetadata(),
+            response.getDeserializedHeaders().getBlobCommittedBlockCount());
+
+        // Write to the file.
+        return FluxUtil.writeFile(data, file,
+            chunkNum * optionsReal.getBlockSize())
+            /*
+            Satisfy the return type. We want to eventually give the user back the headers.
+             */
+            .then(Mono.just(new SimpleResponse<>(response.getRequest(), response.getStatusCode(),
+                response.getHeaders(), properties)));
+    }
+
+    private static long extractTotalBlobLength(String contentRange) {
+        return Long.parseLong(contentRange.split("/")[1]);
+    }
+
+    private void downloadToFileCleanup(AsynchronousFileChannel channel, String filePath, SignalType signalType) {
+        try {
+            channel.close();
+            if (!signalType.equals(SignalType.ON_COMPLETE)) {
```

Comment:
This is a safeguard in case the operation fails and the user won't have a potentially corrupted file, correct? If so, would we want to log a message here stating that the operation wasn't able to successfully complete, cleaning up resources.